src/wallet/transaction.h

```diff
@@ -128,21 +128,35 @@ std::string TxStateString(const T& state)
 }
 
 /**
- * Cachable amount subdivided into watchonly and spendable parts.
+ * Cachable amount subdivided into avoid reuse and all balances
  */
 struct CachableAmount
 {
-    // NO and ALL are never (supposed to be) cached
-    std::bitset<ISMINE_ENUM_ELEMENTS> m_cached;
-    CAmount m_value[ISMINE_ENUM_ELEMENTS];
+    std::optional<CAmount> m_avoid_reuse_value;
+    std::optional<CAmount> m_all_value;
     inline void Reset()
     {
-        m_cached.reset();
+        m_avoid_reuse_value.reset();
+        m_all_value.reset();
     }
-    void Set(isminefilter filter, CAmount value)
+    void Set(bool avoid_reuse, CAmount value)
     {
-        m_cached.set(filter);
-        m_value[filter] = value;
+        if (avoid_reuse) {
+            m_avoid_reuse_value = value;
+        } else {
+            m_all_value = value;
+        }
+    }
+    CAmount Get(bool avoid_reuse)
+    {
+        Assert(m_avoid_reuse_value.has_value() || m_all_value.has_value());
+        if (avoid_reuse) return m_avoid_reuse_value.value();
+        return m_all_value.value();
```

Comment:
This seems unsafe. If we pass `avoid_reuse` as true but only `m_all_value` is set, then `m_avoid_reuse_value.value()` should throw.

Should probably be something like

```
    if (avoid_reuse) {
        Assert(m_avoid_reuse_value.has_value());
        return m_avoid_reuse_value.value();
    }
    Assert(m_all_value.has_value());
    return m_all_value.value();
```

Reply:
Good point, done.